### PR TITLE
TAS: validate the preferred and required annotations for Job and JobSet

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Part of #2724 

#### Special notes for your reviewer:

Only covering Job and JobSet in the PR, but I would like to follow this up extending the support for all Job types hopefully for alpha. If the pattern is established it should be rather fast.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/jobframework/tas_validation.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+func ValidateTASPodSetRequest(replicaPath *field.Path, replicaMetadata *metav1.ObjectMeta) field.ErrorList {
+	var allErrs field.ErrorList
+	requiredValue, requiredFound := replicaMetadata.Annotations[kueuealpha.PodSetRequiredTopologyAnnotation]
+	preferredValue, preferredFound := replicaMetadata.Annotations[kueuealpha.PodSetPreferredTopologyAnnotation]
+	annotationsPath := replicaPath.Child("annotations")
+	if requiredFound && preferredFound {
+		allErrs = append(allErrs, field.Invalid(annotationsPath, field.OmitValueType{},
+			fmt.Sprintf("both %q and %q specified for the replica",
```
Comment (tenzen-y) [inline pkg/controller/jobframework/tas_validation.go:36]:
```suggestion
			fmt.Sprintf("must not contain both %q and %q",
```

Couldn't we clarify what we must not do?

File: pkg/controller/jobframework/tas_validation.go
Comment (tenzen-y) [inline pkg/controller/jobframework/tas_validation.go:1]: TAS has only one commonized validation func. So, wouldn't we consolidate this file to https://github.com/kubernetes-sigs/kueue/blob/9bd0fccfb464aea76bbcdee9824bde0ae4dc73ce/pkg/controller/jobframework/validation.go

**Review by tenzen-y — COMMENTED** [review]
Basically, lgtm

File: pkg/controller/jobframework/tas_validation.go
Comment (mimowo) [reply pkg/controller/jobframework/tas_validation.go:1]: We could, I'm ok either way. 

My thinking was that we may have more TAS-specific validations in the future and then separating will be harder. I also used the tas prefix for other places like cache. WDYT?

File: pkg/controller/jobframework/tas_validation.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/jobframework/tas_validation.go]: done

**Comment by mimowo** [comment]
actually, @tenzen-y wdyt about instead of webhook validation just deactivate the workload in workload controller? it would be much less code and work for free for all integrations?

**Comment by tenzen-y** [comment]
> actually, @tenzen-y wdyt about instead of webhook validation just deactivate the workload in workload controller? it would be much less code and work for free for all integrations? it would be a form of 'soft' validation as we plan for CQs.
> 
> /hold to clarify on the approach, I can prototype it in a separate pr next week

I believe that the validation approach is better than deactivation since, in the case of deactivation, they need to seek the reason why their Job was not admitted.
They do not have knowledge of the TAS limitations that can not be specified in both fields (preferred and required) at the same time. So I guess that they often fall in rabbit hole.

**Comment by mimowo** [comment]
Yeah, I agree it is better from user's perspective to fail fast. I was just thinking it might be less code to validate at the workload level. If you think it is more important pros the user experience then I consider the PR ready to review and merfe

File: pkg/controller/jobframework/tas_validation.go
Comment (tenzen-y) [reply pkg/controller/jobframework/tas_validation.go:1]: Alright. I do not have strong objections to this file.
Let's keep this one.

**Comment by tenzen-y** [comment]
> Yeah, I agree it is better from user's perspective to fail fast. I was just thinking it might be less code to validate at the workload level. If you think it is more important pros the user experience then I consider the PR ready to review and merge

Yeah, your POV also has a point, but I believe that the UX is more important.
So, let me merge this PR.

**Review by tenzen-y — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**